File: app/src/main/java/sixue/naivereader/provider/PtwxProvider.java
package sixue.naivereader.provider;

import android.content.Context;
import android.util.Log;

import org.jsoup.Connection;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.List;

import sixue.naivereader.Utils;
import sixue.naivereader.data.Book;
import sixue.naivereader.data.BookKind;
import sixue.naivereader.data.Chapter;
import sixue.naivereader.data.Source;
import sixue.naivereader.helper.OnlineHelper;

public class PtwxProvider extends NetProvider {

    private static final String TAG = PtwxProvider.class.getSimpleName();

    @Override
    public String getProviderId() {
        return "www.piaotian.net";
    }

    @Override
    public String getProviderName() {
        return "飘天文学";
    }

    @Override
    public List<Book> search(String s, Context context) {
        List<Book> list = new ArrayList<>();
        try {
            String key = URLEncoder.encode(s, "GB2312");
            String url = "https://www.ptwxz.com/modules/article/search.php?searchkey=" + key;
            Connection.Response response = Jsoup.connect(url).followRedirects(true).timeout(5000).execute();
            if (!url.equals(response.url().toString())) {
                Document doc = response.parse();

                Elements elements = doc.body().select("#content");

                Document subDoc = Jsoup.parse(elements.toString());
                String author = "*";
                for (Element td : subDoc.select("td")) {
                    String t = td.text();
                    Log.i(getClass().toString(), "td.text=" + t);

                    final String prefix = "作\u00a0\u00a0\u00a0 者：";
                    if (t.startsWith(prefix)) {
                        author = t.substring(prefix.length());
                        break;
                    }
                }

                String para = parseBookUrl(response.url().toString());
                String coverUrl = calcCoverUrl(para);

                Book book = new Book();
                book.setId(s);
                book.setTitle(s);
                book.setAuthor(author);
                book.setKind(BookKind.Online);

                Source source = new Source();
                source.setId(getProviderId());
                source.setPara(para);
                book.getSources().add(source);

                book.setSiteId(source.getId());
                book.setSitePara(source.getPara());

                OnlineHelper helper = (OnlineHelper) book.buildHelper();
                helper.downloadCover(context, coverUrl);

                list.add(book);
            } else {
                Document doc = response.parse();

                Elements elements = doc.body().select("table.grid");
                for (Element tr : Jsoup.parse(elements.toString()).select("tr")) {
                    Elements tds = tr.select("td.odd");
                    Elements a = tds.select("a");
                    if (tds.size() == 0) {
                        continue;
                    }

                    String para = parseBookUrl(a.attr("href").trim());
                    String coverUrl = calcCoverUrl(para);

                    Book book = new Book();

                    String title = a.text();
                    String id = a.text();
                    String author = tds.get(1).text();

                    book.setId(id);
                    book.setTitle(title);
                    book.setAuthor(author);
                    book.setKind(BookKind.Online);

                    Source source = new Source();
                    source.setId(getProviderId());
                    source.setPara(para);
                    book.getSources().add(source);

                    book.setSiteId(source.getId());
                    book.setSitePara(source.getPara());

                    OnlineHelper helper = (OnlineHelper) book.buildHelper();
                    helper.downloadCover(context, coverUrl);

                    list.add(book);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return list;
    }

    @Override
    public List<Chapter> downloadContent(Book book, String bookSavePath) {

        String contentUrl = calcBookUrl(book.getSitePara());
        List<Chapter> content = new ArrayList<>();
        try {
            Document doc = Jsoup.connect(contentUrl).timeout(5000).get();
            Elements elements = doc.body().select(".centent");
            for (Element ch : Jsoup.parse(elements.toString()).select("li")) {
                String title = ch.select("a").text();
                String url = ch.select("a").attr("href").replace("/", "").trim();
                if (url.length() == 0 ||
                        url.toLowerCase().startsWith("javascript:") ||
                        url.toLowerCase().startsWith("window")) {
                    continue;
                }

                final Chapter chapter = new Chapter();
                chapter.setId(url);
                chapter.setTitle(title);
                String chapterSavePath = calcChapterSavePath(chapter, bookSavePath);
                chapter.setSavePath(chapterSavePath);

                content.add(chapter);
            }
        } catch (IOException | NullPointerException e) {
            Log.e(TAG, "downloadContent ERROR: " + contentUrl);
        }
        return content;
    }

    @Override
    public void downloadChapter(Book book, Chapter chapter) {

        String chapterUrl = getChapterUrl(book, chapter);
        try {
            Document doc = Jsoup.connect(chapterUrl).timeout(5000).get();
            String s = doc.body().toString()
                    .replace("<script language=\"javascript\">GetFont();</script>", "<div id=\"content\" class=\"fonts_mesne\">")
                    .replace("<!-- 翻页上AD开始 -->", "</div> <!-- 翻页上AD开始 -->");
            Element content = Jsoup.parse(s).select("#content").first();
            String text = content.html();
            String plainText = Utils.clearHtmlTag(text, new String[] {"h1", "table", "div"})
                    .replace("<br>", "")
                    .replace("&nbsp;", " ");
            Utils.writeText(plainText, chapter.getSavePath());
        } catch (IOException | NullPointerException e) {
            Log.e(TAG, "downloadChapter ERROR: " + chapterUrl);
        }
    }

    @Override
    public String getChapterUrl(Book book, Chapter chapter) {
        return calcBookUrl(book.getSitePara()) + "/" + chapter.getId();
    }

    private String calcChapterSavePath(Chapter chapter, String bookSavePath) {
        return bookSavePath + "/" + chapter.getId().replace(".html", ".txt");
    }

    private String calcPrefix(String para) {
        return para.length() > 3 ? para.substring(0, para.length() - 3) : "0";
    }

    private String calcBookUrl(String para) {
        String prefix = calcPrefix(para);
        return String.format("https://www.ptwxz.com/html/%s/%s", prefix, para);
    }

    private String calcCoverUrl(String para) {
        String prefix = calcPrefix(para);
        return String.format("https://www.ptwxz.com/files/article/image/%s/%s/%ss.jpg", prefix, para, para);
    }

    private String parseBookUrl(String bookUrl) {
        int l = bookUrl.lastIndexOf('/');
        int r = bookUrl.lastIndexOf(".html");
        if (l != -1 && r != -1 && l < r) {
            return bookUrl.substring(l + 1, r);
        } else {
            return "";
        }
    }
}

File: app/src/main/java/sixue/naivereader/provider/PtwxProvider.kt
package sixue.naivereader.provider

import android.content.Context
import android.util.Log
import org.jsoup.Jsoup
import sixue.naivereader.Utils
import sixue.naivereader.data.Book
import sixue.naivereader.data.BookKind
import sixue.naivereader.data.Chapter
import sixue.naivereader.data.Source
import sixue.naivereader.helper.OnlineHelper
import java.io.IOException
import java.net.URLEncoder
import java.util.*

class PtwxProvider : NetProvider() {
    override val providerId: String
        get() = "www.piaotian.net"
    override val providerName: String
        get() = "飘天文学"

    override fun search(s: String, context: Context): List<Book> {
        val list: MutableList<Book> = ArrayList()
        try {
            val key = URLEncoder.encode(s, "GBK")
            val url = "https://www.piaotia.com/modules/article/search.php?searchkey=$key"
            val response = Jsoup.connect(url).followRedirects(true).timeout(15000).execute()
            if (url != response.url().toString()) {
                val doc = response.parse()
                val elements = doc.body().select("#content")
                val subDoc = Jsoup.parse(elements.toString())
                var author = "*"
                for (td in subDoc.select("td")) {
                    val t = td.text()
                    Log.i(javaClass.toString(), "td.text=$t")
                    val prefix = "作 者："
                    if (t.startsWith(prefix)) {
                        author = t.substring(prefix.length)
                        break
                    }
                }
                val para = parseBookUrl(response.url().toString())
                val coverUrl = calcCoverUrl(para)
                val book = Book(
                    id = s,
                    title = s,
                    author = author,
                    kind = BookKind.Online,
                    localPath = "",
                )
                val source = Source(
                    id = providerId,
                    para = para,
                )
                book.sources += source
                book.siteId = source.id
                book.sitePara = source.para
                val helper = book.buildHelper() as OnlineHelper
                helper.downloadCover(context, coverUrl)
                list.add(book)
            } else {
                val doc = response.parse()
                val elements = doc.body().select("table.grid")
                for (tr in Jsoup.parse(elements.toString()).select("tr")) {
                    val tds = tr.select("td.odd")
                    val a = tds.select("a")
                    if (tds.size == 0) {
                        continue
                    }
                    val para = parseBookUrl(a.attr("href").trim { it <= ' ' })
                    val coverUrl = calcCoverUrl(para)
                    val title = a.text()
                    val id = a.text()
                    val author = tds[1].text()
                    val book = Book(
                        id = id,
                        title = title,
                        author = author,
                        kind = BookKind.Online,
                        localPath = "",
                    )
                    val source = Source(
                        id = providerId,
                        para = para
                    )
                    book.sources += source
                    book.siteId = source.id
                    book.sitePara = source.para
                    val helper = book.buildHelper() as OnlineHelper
                    helper.downloadCover(context, coverUrl)
                    list.add(book)
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return list
    }

    override fun parseBookUrl(s: String, context: Context): String? {
        return null
    }

    override fun loadBookUrl(s: String, context: Context): Book? {
        return null
    }

    override fun downloadContent(book: Book, bookSavePath: String): List<Chapter> {
        val contentUrl = calcBookUrl(book.sitePara!!)
        val content: MutableList<Chapter> = ArrayList()
        try {
            val doc = Jsoup.connect(contentUrl).timeout(5000).get()
            val elements = doc.body().select(".centent")
            for (ch in Jsoup.parse(elements.toString()).select("li")) {
                val title = ch.select("a").text()
                val url = ch.select("a").attr("href").replace("/", "").trim { it <= ' ' }
                if (url.isEmpty() ||
                        url.lowercase(Locale.getDefault()).startsWith("javascript:") ||
                        url.lowercase(Locale.getDefault()).startsWith("window")) {
                    continue
                }
                val chapter = Chapter(
                    id = url,
                    title = title,
                )
                val chapterSavePath = calcChapterSavePath(chapter, bookSavePath)
                chapter.savePath = chapterSavePath
                content.add(chapter)
            }
        } catch (e: IOException) {
            Log.e(TAG, "downloadContent ERROR: $contentUrl")
        } catch (e: NullPointerException) {
            Log.e(TAG, "downloadContent ERROR: $contentUrl")
        }
        return content
    }

    override fun downloadChapter(book: Book, chapter: Chapter) {
        val chapterUrl = getChapterUrl(book, chapter)
        try {
            val doc = Jsoup.connect(chapterUrl).timeout(5000).get()
            val s = doc.body().toString()
                    .replace("<script language=\"javascript\">GetFont();</script>", "<div id=\"content\" class=\"fonts_mesne\">")
                    .replace("<!-- 翻页上AD开始 -->", "</div> <!-- 翻页上AD开始 -->")
            val content = Jsoup.parse(s).select("#content").first()!!
            val text = content.html()
            val plainText = Utils.clearHtmlTag(text, arrayOf("h1", "table", "div"))
                    .replace("<br>", "")
                    .replace("&nbsp;", " ")
            Utils.writeText(plainText, chapter.savePath)
        } catch (e: IOException) {
            Log.e(TAG, "downloadChapter ERROR: $chapterUrl")
        } catch (e: NullPointerException) {
            Log.e(TAG, "downloadChapter ERROR: $chapterUrl")
        }
    }

    override fun getChapterUrl(book: Book, chapter: Chapter): String {
        return calcBookUrl(book.sitePara!!) + "/" + chapter.id
    }

    private fun calcChapterSavePath(chapter: Chapter, bookSavePath: String): String {
        return bookSavePath + "/" + chapter.id.replace(".html", ".txt")
    }

    private fun calcPrefix(para: String): String {
        return if (para.length > 3) para.substring(0, para.length - 3) else "0"
    }

    private fun calcBookUrl(para: String): String {
        val prefix = calcPrefix(para)
        return String.format("https://www.piaotia.com/html/%s/%s", prefix, para)
    }

    private fun calcCoverUrl(para: String): String {
        val prefix = calcPrefix(para)
        return String.format("https://www.piaotia.com/files/article/image/%s/%s/%ss.jpg", prefix, para, para)
    }

    private fun parseBookUrl(bookUrl: String): String {
        val l = bookUrl.lastIndexOf('/')
        val r = bookUrl.lastIndexOf(".html")
        return if (l != -1 && r != -1 && l < r) {
            bookUrl.substring(l + 1, r)
        } else {
            ""
        }
    }

    companion object {
        private val TAG = PtwxProvider::class.java.simpleName
    }
}